### PR TITLE
Only Add NullHandler to Module-Root Logger

### PR DESCRIPTION
**Pull request by gnossen**
I was trying to get a feel for what the rest of the python ecosystem does with its logging, so I looked into the top few libraries on pypi:

### `urllib3`

`urllib3` maintains a logger for not quite every module, but for each one that does heavy lifting. The logger name is `__name__`, and no handlers are registered for any module-level loggers, including `NullHandler`. Their documentation [spells out how to configure logging for the library](https://github.com/urllib3/urllib3/blob/0f85e05af9ef2ded671a7b47506dfd24b32decf0/docs/user-guide.rst#logging).

They explicitly configure a library root-level logger called `urllib3` to which they attach a `NullHandler`. This addresses the "no handlers could be found" problem.

Their tests explicitly configure handlers, just like ours do.

### `scrapy`

`scrapy` is more hands-on. It provides a [configuration module for its logging](https://github.com/scrapy/scrapy/blob/master/scrapy/utils/log.py) and [a whole document](https://github.com/scrapy/scrapy/blob/master/docs/topics/logging.rst) on how to handle logging with `scrapy`. It looks like `log.py`'s whole reason for existence is making sure that a handler is attached to to the `scrapy` handler at startup.

I think the extra complexity here is because `scrapy` also offers a CLI, so there has to be some way to configure logging without resorting to writing python.

---
Based on all of the libraries I've looked at, I think our current approach is reasonable. The one change I would make is to explicitly configure a `grpc` logger and to only attach a `NullHandler` to it instead of putting the burden on the author of each new module to configure it there.

### tl;dr
I think we should
 - Configure a logger in each module that cares about logging
 - Do *not* attach any handler to the module level loggers (including `NullHandler`s)
 - Explicity configure a `grpc` logger with a `NullHandler` attached

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,390      Total (=)      2,019,390

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,180,038      Total (>)     11,180,035

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,390      Total (=)      2,019,390

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,180,034      Total (<)     11,180,035

 No significant differences in binary sizes


```

**Comment by gnossen** [comment]
Known flakes:
 - https://github.com/grpc/grpc/issues/17138

File: src/python/grpcio_tests/tests/unit/_logging_test.py [hunk 1]
```diff
@@ -68,6 +68,13 @@ def test_can_configure_logger(self):
         self.assertEqual(1, len(logging.getLogger().handlers))
         self.assertIs(logging.getLogger().handlers[0].stream, intended_stream)
 
+    @isolated_logging
+    def test_grpc_logger(self):
+        self.assertIn("grpc", logging.Logger.manager.loggerDict.keys())
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/unit/_logging_test.py]: Why `.keys()` on this line?

Comment (nathanielmanistaatgoogle) [reply src/python/grpcio_tests/tests/unit/_logging_test.py]: GitHub Code Review is not yet showing me the fix - have you yet to repush?

**Comment by gnossen** [comment]
@nathanielmanistaatgoogle Sorry about that. I pushed said changes and formatted the PR as requested. I've not used the squash model in the past. Since it requires force pushing, is your preference to only squash once all comments have been addressed or to squash as you go?

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,040      Total (=)      2,019,040

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,179,737      Total (>)     11,179,727

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,040      Total (=)      2,019,040

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,179,736      Total (<)     11,179,741

 No significant differences in binary sizes


```

**Comment by gnossen** [comment]
@nathanielmanistaatgoogle Squashed.

**Comment by gnossen** [comment]
Known failure: https://github.com/grpc/grpc/issues/17170

Should be fixed on master already.
